src/runtime_src/core/pcie/emulation/common_em/config.h

```diff
@@ -89,6 +90,89 @@ namespace xclemulation{
   const int VIVADO_MIN_VERSION = 2000;
   const int VIVADO_MAX_VERSION = 2100;
 
+inline bool file_exists(const std::string& fnm)
+{
+  struct stat statBuf;
+  return stat(fnm.c_str(), &statBuf) == 0;
+}
+
+
+//sparse log utility
+enum class eEmulationType{
+  eSw_Emu,
+  eHw_Emu
+};
+
+struct sParseLog
+{
+  std::ifstream file;
+  std::string mFileName;
+  std::atomic_bool mFileExists;
+  std::vector<std::string> mMatchedStrings;
+  eEmulationType mEmuType;
+
+  sParseLog(const std::string &iDeviceLog, eEmulationType iType, std::vector<std::string> &iMatchedStrings)
+      : mFileName(iDeviceLog), mMatchedStrings(iMatchedStrings), mEmuType(iType)
+  {
+
+    mFileExists.store(false);
+  }
+
+  ~sParseLog() = default;
+
+  void SearchString()
+  {
+    std::string line;
+    while (std::getline(file, line))
+    {
+      for (const auto &StringData : mMatchedStrings)
+      {
+        if (line.find(StringData) != std::string::npos)
+        {
+          if (eEmulationType::eSw_Emu == mEmuType)
+            printMsg();
+          else if (eEmulationType::eHw_Emu == mEmuType)
+          {
+            if (!StringData.compare("Exiting xsim") || !StringData.compare("FATAL_ERROR"))
+              printMsg();
+            else
+              std::cout << line << std::endl;
+          }
+        }
+      }
+    }
+  }
+  void printMsg()
+  {
+    switch (mEmuType)
+    {
+    case eEmulationType::eSw_Emu:
+      std::cout << "Received request to end the application. Press Cntrl+C to exit the application." << std::endl;
+      break;
+    case eEmulationType::eHw_Emu:
+      std::cout << "SIMULATION EXITED" << std::endl;
```

Comment:
Here std::endl makes sense.
But I don't understand the SearchString() function and how possible printMsg() can be called multiple times.

The naming convention is horrid, some functions are UpperCase() some are lowerCase(),  IMO they should all be lower_case().

Reply:
Newly added description of HwEmShim::messagesThread() (debug.cxx file) explains the possible number of calls to  sparseLog which indirectly calls this printMsg().

All methods in this structure are following lower_case now.